From stable-bounces@linux.kernel.org Mon Sep 25 16:25:34 2006
Message-Id: <200609252324.k8PNOeYQ002067@shell0.pdx.osdl.net>
To: torvalds@osdl.org
From: keith mannthey <kmannth@us.ibm.com>
Date: Mon, 25 Sep 2006 16:24:39 -0700
Cc: akpm@osdl.org, haveblue@us.ibm.com, bunk@stusta.de, vgoyal@in.ibm.com,
        kmannth@us.ibm.com, stable@kernel.org
Subject: i386 bootioremap / kexec fix

From: keith mannthey <kmannth@us.ibm.com>

With CONFIG_PHYSICAL_START set to a non default values the i386
boot_ioremap code calculated its pte index wrong and users of boot_ioremap
have their areas incorrectly mapped (for me SRAT table not mapped during
early boot).  This patch removes the addr < BOOT_PTE_PTRS constraint.

Signed-off-by: Keith Mannthey<kmannth@us.ibm.com>
Cc: Vivek Goyal <vgoyal@in.ibm.com>
Cc: Dave Hansen <haveblue@us.ibm.com>
Cc: Adrian Bunk <bunk@stusta.de>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 arch/i386/mm/boot_ioremap.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- linux-2.6.18.orig/arch/i386/mm/boot_ioremap.c
+++ linux-2.6.18/arch/i386/mm/boot_ioremap.c
@@ -29,8 +29,11 @@
  */
 
 #define BOOT_PTE_PTRS (PTRS_PER_PTE*2)
-#define boot_pte_index(address) \
-	     (((address) >> PAGE_SHIFT) & (BOOT_PTE_PTRS - 1))
+
+static unsigned long boot_pte_index(unsigned long vaddr)
+{
+	return __pa(vaddr) >> PAGE_SHIFT;
+}
 
 static inline boot_pte_t* boot_vaddr_to_pte(void *address)
 {
